Change node-role.kubernetes.io from master to control-plane

floryut opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Change label from master to control-plane:
- Kubeadm: http://git.k8s.io/enhancements/keps/sig-cluster-lifecycle/kubeadm/2067-rename-master-label-taint/README.md (https://github.com/kubernetes/kubernetes/pull/95382, @neolit123)

**Which issue(s) this PR fixes**:
none

**Special notes for your reviewer**:
none

**Does this PR introduce a user-facing change?**:
```release-note
`node-role.kubernetes.io/master` is now rename to `node-role.kubernetes.io/control-plane`
```

neolit123 reviewed (commented):
one problem with a hard rename is that that kubespray user's scripts that depend on the label/taints get broken when this version releases.

did kubespray announce this as deprecated in advance?

neolit123 commented on roles/kubernetes/master/templates/kubeadm-config.v1beta2.yaml.j2:
```diff
@@ -19,7 +19,7 @@ nodeRegistration:
 {% if inventory_hostname in groups['kube-master'] and inventory_hostname not in groups['kube-node'] %}
```
should kubespray phase away from using the master term in general?

floryut replied:
Pretty much yeah, we have something in our pipe but it's a big change :)

neolit123 commented on roles/kubernetes/master/templates/kubeadm-config.v1beta2.yaml.j2:
```diff
@@ -19,7 +19,7 @@ nodeRegistration:
   taints:
   - effect: NoSchedule
-    key: node-role.kubernetes.io/master
+    key: node-role.kubernetes.io/control-plane
```
noting that existing user workloads that do not tolerate this new taint will fail to schedule.

floryut replied:
Yup that's why I flag the PR as draft, I have some testing to do and will probably only change label for now

floryut commented:
> one problem with a hard rename is that that kubespray user's scripts that depend on the label/taints get broken when this version releases.
> 
> did kubespray announce this as deprecated in advance?

Ok I've update to follow kubernetes guideline.:
> Renaming the "node-role.kubernetes.io/master" Node label => Announce to users that they should adapt to use the new label.
I've change to use control-plane label instead of master.
> Renaming the "node-role.kubernetes.io/master" Node taint => Announce to users that they should do that same for their workloads.
I've added the toleration but I kept the old one as the control-plane taint will be added to node in a future release.

LuckySB commented:
From which version will the kubeadm assign a label 'control-plane' to the master nodes ?

Will it change the label `master` to `control-plane` when updating the cluster ?

LuckySB commented on roles/kubernetes-apps/cloud_controller/oci/templates/oci-cloud-provider.yml.j2:
```diff
@@ -36,14 +36,17 @@ spec:
       dnsPolicy: ClusterFirstWithHostNet
       nodeSelector:
-        node-role.kubernetes.io/master: ""
+        node-role.kubernetes.io/control-plane: ""
```
this change breake launch this pods, if no node has this labels.

need to change `nodeSelector:` to NodeAffinity with multiple labels

LuckySB replied:
or need check kube_version and use new labels only if version greate then 1.20

floryut replied:
true

neolit123 commented:
> From which version will the kubeadm assign a label 'control-plane' to the master nodes ?

it started in 1.20.

neolit123 commented:
> I've added the toleration but I kept the old one as the control-plane taint will be added to node in a future release.
> WDYT ?

the change has to be carefully planned.

probably better to only add the tolerations now and add the taints only on a later stage.

it also depends on what kubeadm versions this version of kubespray can deploy.
the control-plane label is only present in 1.20+.

this is outlined in the linked KEP.

floryut commented:
> > From which version will the kubeadm assign a label 'control-plane' to the master nodes ?
> 
> it started in 1.20.

Yup starting from 1.20 : `Introduce the "node-role.kubernetes.io/control-plane" label in parallel to the "master" label.`
Then in a later version `Remove the "master" label and announce it to the users.`

floryut commented:
@LuckySB Ok, I've added check for k8s version, to enable label selection on `control-plane` only for 1.20+
The `control-plane` toleration is also added, we'll only need to cleanup both later (well way way later :D)

LuckySB commented on roles/kubernetes-apps/csi_driver/azuredisk/templates/azure-csi-azuredisk-controller.yml.j2:
```diff
@@ -24,6 +24,10 @@ spec:
           operator: "Equal"
           value: "true"
```
remove `value: true` and `operator: Equal`

on modern cluster value is `""`, no `"true"`

floryut replied:
done

LuckySB reviewed (commented):
I completely agree with @neolit123 

Adding this taint  at this moment may can break the launch of others third party softs.
for example - csi drivers, monitor agents, log collectors started by daemonsets

New taint planned to add only in [second stage](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/kubeadm/2067-rename-master-label-taint/README.md#renaming-the-node-rolekubernetesiomaster-node-taint)
in next year

floryut commented:
> I completely agree with @neolit123
> 
> Adding this taint at this moment may can break the launch of others third party softs.
> for example - csi drivers, monitor agents, log collectors started by daemonsets
> 
> New taint planned to add only in [second stage](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/kubeadm/2067-rename-master-label-taint/README.md#renaming-the-node-rolekubernetesiomaster-node-taint)
> in next year

Well  yeah, but adding the taint or not is not up to us but kubeadm.
In this PR i'm now only adding the TOLERATION, as kubernetes recommend and are doing so with CoreDNS in parralel with master toleration and renoming the label for k8s 1.20.

LuckySB commented:
in this code
https://github.com/kubernetes-sigs/kubespray/pull/7183/files#diff-0be1b4e4a9316f81ccc4a70601c1f3bfbf40b85637da5507048b5f7c4fd3baf3R15-R27

you ADD second TAINT to node in kube-master group 
```
nodeRegistration:
   taints:
```

floryut commented:
> in this code
> 
> https://github.com/floryut/kubespray/blob/cca313cbb47f55a98dc38f8bf651804a57501eac/roles/kubernetes/master/templates/kubeadm-config.v1beta2.yaml.j2#L15-L27
> 
> ```
> >nodeRegistration:
> {% if inventory_hostname in groups['kube-master'] and inventory_hostname not in groups['kube-node'] %}
> >  taints:
>   - effect: NoSchedule
>     key: node-role.kubernetes.io/master
>   - effect: NoSchedule
>     key: node-role.kubernetes.io/control-plane
> {% else %}
>   taints: []
> {% endif %}
> ```
> 
> you ADD second TAINT to node in kube-master group
> 
> ```
> nodeRegistration:
>    taints:
> ```

Oh f** mistake on my end! Will update thanks @LuckySB

oomichi commented on roles/kubernetes-apps/ansible/templates/dns-autoscaler.yml.j2:
```diff
@@ -40,8 +40,9 @@ spec:
         kubernetes.io/os: linux
       tolerations:
         - effect: NoSchedule
-          operator: Equal
```
The reason why this line is removed is the default value of operator is `Equal` and most other manifests don't specify it?

floryut replied:
Yup default value is `Equal` for operator

oomichi reviewed (commented):
confirmed all parts which contain `node-role.kubernetes.io/master` are covered by this PR.
Just a small question.

/cc @oomichi

floryut commented:
> I completely agree with @neolit123
> 
> Adding this taint at this moment may can break the launch of others third party softs.
> for example - csi drivers, monitor agents, log collectors started by daemonsets
> 
> New taint planned to add only in [second stage](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/kubeadm/2067-rename-master-label-taint/README.md#renaming-the-node-rolekubernetesiomaster-node-taint)
> in next year

ok done

LuckySB commented:
/lgtm
